src/lib_file_deps.ml

```diff
@@ -1,79 +1,51 @@
-      |> String.concat  ~sep:"-and-"
-
-    let alias t ~dir ~name =
-      sprintf "lib-%s%s-all" (Lib_name.to_string name) (to_string t)
-      |> Alias.make ~dir
-
-    let setup_alias t ~dir ~lib ~files =
-      Build_system.Alias.add_deps
-        (alias t ~dir ~name:(Library.best_name lib))
-        files
-
-    let setup_file_deps_group_alias t ~dir ~lib =
-      setup_alias t ~dir ~lib ~files:(
-        List.map t ~f:(fun t ->
-          Alias.stamp_file (alias [t] ~dir ~name:(Library.best_name lib)))
-        |> Path.Set.of_list
-      )
-  end
-end
-
-let setup_file_deps =
-  let cm_kinds = [Cm_kind.Cmx; Cmi] in
-  let groups = List.map ~f:Group.of_cm_kind cm_kinds in
-  fun ~dir ~lib ~modules ->
-    let add_cms ~cm_kind =
-      List.fold_left ~f:(fun acc m ->
-        match Module.cm_public_file m cm_kind with
-        | None -> acc
-        | Some fn -> Path.Set.add acc fn)
+  let obj_dirs t obj_dir =
+    match t with
+    | Cmi -> [Obj_dir.byte_dir obj_dir; Obj_dir.public_cmi_dir obj_dir]
+    | Cmx -> [Obj_dir.native_dir obj_dir]
+    | Header -> [Obj_dir.dir obj_dir]
+
+  let to_predicate =
+    let preds = List.map all ~f:(fun g ->
+      let ext = to_string g in
+      (* we cannot use globs because of bootstrapping. But we reuse the same
+         repr just in case the user uses this glob *)
+      let id = lazy (Glob.id (sprintf "*%s" ext)) in
+      let pred = Predicate.create ~id ~f:(fun p ->
+        String.equal (Path.extension p) ext)
+      in
+      (g, pred))
     in
-    List.iter cm_kinds ~f:(fun cm_kind ->
-      let files = add_cms ~cm_kind ~init:Path.Set.empty modules in
-      let groups = [Group.of_cm_kind cm_kind] in
-      Group.L.setup_alias groups ~dir ~lib ~files);
-    Group.L.setup_file_deps_group_alias groups ~dir ~lib;
-    Group.L.setup_alias [Header] ~dir ~lib ~files:(
-      List.map lib.install_c_headers ~f:(fun header ->
-        Path.relative dir (header ^ ".h"))
-      |> Path.Set.of_list)
-
-let file_deps_of_lib (lib : Lib.t) ~groups =
-  if Lib.is_local lib then
-    [Alias.stamp_file
-       (Group.L.alias groups ~dir:(Lib.src_dir lib) ~name:(Lib.name lib))]
-  else
-    (* suppose that all the files of an external lib are at the same place *)
-    Build_system.stamp_files_for_files_of
-      ~dir:(Obj_dir.public_cmi_dir (Lib.obj_dir lib))
-      ~exts:(List.map ~f:Group.to_string groups)
-
-let file_deps_with_exts =
-  List.concat_map ~f:(fun (lib, groups) -> file_deps_of_lib lib ~groups)
+    fun g -> Option.value_exn (List.assoc preds g)
+end
 
-let file_deps libs ~groups =
-  List.concat_map libs ~f:(file_deps_of_lib ~groups)
+let deps_of_lib (lib : Lib.t) ~groups =
+  let obj_dir = Lib.obj_dir lib in
+  List.concat_map groups ~f:(fun g ->
+    List.map (Group.obj_dirs g obj_dir) ~f:(fun dir ->
+      Dep.glob ~dir (Group.to_predicate g)))
+  |> Dep.Set.of_list
+
+let deps_with_exts =
+  List.fold_left ~init:Dep.Set.empty ~f:(fun acc (lib, groups) ->
```

Comment:
I changed this to get rid of `fold` here because I think fold is pretty hard to read and think about. We now have `Set.union_all` and `Set.union_map`.

Reply:
Fine with me.